finesse/gui/measure_script/script.py

```diff
@@ -124,3 +123,235 @@ def parse_script(script: Union[str, TextIOBase]) -> Dict[str, Any]:
+            self._on_stepper_motor_error, f"serial.{STEPPER_MOTOR_TOPIC}.error"
+        )
+
+        # Listen for EM27 messages
+        pub.subscribe(self._measuring_error, "opus.error")
+        pub.subscribe(self._measuring_started, "opus.response.start")
+        pub.subscribe(self._status_received, "opus.response.status")
+
+    def _load_next_measurement(self) -> bool:
+        """Load the next measurement in the sequence.
+
+        Returns:
+            False if there are no more measurements
+        """
+        try:
+            self.current_measurement = next(self.measurement_iter)
+            self.current_measurement_count = 0
+            return True
+        except StopIteration:
+            return False
+
+    def on_enter_moving(self) -> None:
+        """Try to load the next measurement and start the next movement.
+
+        If there are no more measurements, the ScriptRunner will return to a not_running
+        state.
+        """
+        if not self._load_next_measurement():
+            self.finish()
+            return
+
+        # Start moving the stepper motor
+        pub.sendMessage(
+            f"serial.{STEPPER_MOTOR_TOPIC}.move.begin",
+            target=self.current_measurement.angle,
+        )
+
+        # Flag that we want a message when the movement has stopped
+        pub.sendMessage(f"serial.{STEPPER_MOTOR_TOPIC}.notify_on_stopped")
+
+    def on_enter_measuring(self) -> None:
+        """Tell the EM27 to start a new measurement.
+
+        NB: This is also invoked on repeat measurements
+        """
+        pub.sendMessage("opus.request", command="start")
+
+    def on_exit_measuring(self) -> None:
+        """Ensure that the polling timer is stopped."""
+        self._measure_poll_timer.stop()
+
+    def _measuring_started(
+        self,
+        status: int,
+        text: str,
+        error: Optional[tuple[int, str]],
+        url: str,
+    ):
+        """Start polling the EM27 so we know when the measurement is finished."""
+        if error:
+            self._on_em27_error_message(*error)
+        else:
+            _poll_em27_status()
+
+    def _status_received(
+        self,
+        status: int,
+        text: str,
+        error: Optional[tuple[int, str]],
+        url: str,
+    ):
+        """Move on to the next measurement if the measurement has finished."""
+        if error:
+            self._on_em27_error_message(*error)
+            return
+
+        if status == 2:  # "connected" state, indicating measurement is finished
+            self._measuring_end()
+        else:
+            # Poll again later
+            self._measure_poll_timer.start()
+
+    def abort(self) -> None:
+        """Abort the current measure script run."""
+        state = self.current_state
+        if state == self.moving:
+            self.cancel_move()
+        elif state == self.measuring:
+            self.cancel_measuring()
+
+    def _on_stepper_motor_error(self, error: BaseException) -> None:
```

Comment:
My plan was that stepper motor error messages would also trigger an error message giving the user a more informative message about what has gone wrong, though I haven't implemented that yet (and I should open an issue for it). We only care about errors here because we want the measure script to stop executing in this case.